flink-connectors/flink-connector-kafka-0.8/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/SimpleConsumerThread.java

```diff
@@ -373,16 +370,28 @@ else if (partitionsRemoved) {
 								keyPayload.get(keyBytes);
 							}
 
-							final T value = deserializer.deserialize(keyBytes, valueBytes, 
-									currentPartition.getTopic(), currentPartition.getPartition(), offset);
-							
-							if (deserializer.isEndOfStream(value)) {
-								// remove partition from subscribed partitions.
-								partitionsIterator.remove();
-								continue partitionsLoop;
-							}
-							
-							owner.emitRecord(value, currentPartition, offset);
+							final Collector<T> collector = new Collector<T>() {
```

Comment:
@haohui, if you don't mind, I would also wait for @rmetzger to take another look at the new proposals here, before you jump back again into the code.
This part is quite critical for Flink Kafka's exacty-once guarantee, so another pair of eyes on this will be safer.

I would also like to do a thorough pass on your code and see if there are other problems, so you work on those all-together.

Is that ok for you? Sorry for some more waiting.